### PR TITLE
[FEAT] Gitpod Script

### LINKED ISSUE #186: [TODO]  Update gitpod Yaml
With the new all-in-one script from @Dnouv  -  the gitpod deployment is badly outdated and no longer works.

Need to get it updated to support volta based nodeJS/npm version declaration as well as flauna as superprofile in a docker container.

### PR DESCRIPTION
**Pull request by Palanikannan1437**
Closes #186 

## Issue in brief
Create an optimised Gitpod script for the project so that contributors and community builders can enjoy **bam boom boom kapow 🤯!!** speeds to setup and work on the codebase⚡️!

## Suggested Fixes/Changes 

- Added init tasks in `.gitpod.yml` and separated tasks so that pre-builds can be run efficiently
- Tasks that run during initialisation (`init`) for caching via pre-builds
  - Install Faunadb: Pulls fauna image from docker hub so that the results can be cached until next commit(pre-builds)
  - Init Next.js App: Install all npm packages for the Next.js app in the `app` directory
  - Init Strapi App: Install all npm packages for the Strapi app in the `cms` directory and build for production
- Tasks that run during runtime(`command`)
  - Init Strapi App: Starts the Stapi service
  - Start build: Starts the Next.js app once the Strapi app starts
-  `init` tasks are performed asynchronously for max optimization and then the `command` tasks are awaited based on the `init tasks`
- Extracted the initialisation logic from the main build script into `strapi.sh` only for the gitpod-scripts for faster async loading!
- Extracted and modified necessary existing scripts into `gitpod-scripts` 
- Fixed the bug of Next.js app starting before the Strapi service by awaiting for the Strapi port to start listening!
- Updated `Readme` for instructions on how to use pre-builds in Gitpod for their forked version!
- Committed updated `package-lock.json` for keeping this package-lock.json as a baseline for further development to avoid conflicts while initialising gitpod workspace

## Demo
The magical dev-workflow with pre-builds on Gitpod enabled from a new contributor/community builder's perspective!

https://user-images.githubusercontent.com/73993394/203915578-cb30ec31-738f-4767-a858-dfb17c8dda33.mov

PS: These number of commits were made just in-order to test the immediate changes I make on Gitpod, I'll consider squashing them if needed, thank you😄

### CODE REVIEW

**Comment by Sing-Li** [comment]
Looking good @Palanikannan1437 - can't wait to try this!

**Comment by Palanikannan1437** [comment]
> Looking good @Palanikannan1437 - can't wait to try this!

@Sing-Li and @Dnouv it'd be awesome if you could review this and see it in action ⚡️

Please let me know incase there are some changes to be made, thank you 😄

**Comment by Dnouv** [comment]
I am on it. I just noticed a bug in that the `RC4Community` database is not getting initialized. Maybe we can fix it by tweaking the startup shell script; I will take a look at the script.

Thank you!

**Comment by Palanikannan1437** [comment]
> I am on it. I just noticed a bug in that the `RC4Community` database is not getting initialized. Maybe we can fix it by tweaking the startup shell script; I will take a look at the script.
> 
> Update: The following is the error for every new workspace @Palanikannan1437
> 
> ```
> Waiting 30 seconds for container to get shipped...
> Docker container needs extra startup time, please increase the $waittlist value in initFaunaOnce.sh
> Process ended with health status of Container: starting
> ```
> 
> While building on Gitpod; it takes an unusually long time for the Fauna container to start. This results in an empty database, thus skipping the data initialization.
> 
> Thank you!

@Dnouv I've now fixed the script to dynamically increase the waittime by 5s until the container starts...with an upper limit of 60s, would be awesome if you could test it out now, thank you so much for the guidance! 😄

File: superprofile/initFaunaOnce.sh [hunk 1]
```diff
@@ -17,20 +20,33 @@ sleep $waittime
...
-        echo "Copying over secrets to ../app/.env"
-        printf '\nNEXT_PUBLIC_FAUNA_SECRET=' | cat - ./log/dbkey >> ../app/.env &&
-        printf '\nNEXT_PUBLIC_FAUNA_DOMAIN'="http://localhost:8084/graphql" >> ../app/.env
-        touch $DBF &&
-        echo "-- All set, superprofile launch 🚀"
+
+check_and_start_fauna_container() {
+    container_state="$( docker inspect -f '{{ .State.Health.Status }}' ${container_name} )"
+
+    if [ "$container_state" != $healthy ]; then
```
Comment (Dnouv) [inline superprofile/initFaunaOnce.sh:27]: I think it would be better if we check for `container_health==starting` since it will not be efficient to go on increasing wait-time for the container if the health becomes `unhealthy`.

```
health_status = "unhealthy"

// current logic
if (health_status != "healthy") {
             // evaluates to true
             increase_waittime() //unnecessary wait
 }

// updated logic
if (health_status == "starting") {
             // evaluates to false
             // do not run the below steps and quit
            increase_waittime()
 }
```

The health status can be one of the following:
**starting**  – Initial status when the container is still starting
**healthy**   – If the command succeeds then the container is healthy
**unhealthy** – If a single run of the  takes longer than the specified 
  timeout then it is considered unhealthy. If a health check fails then the 
  will run retries number of times and will be declared unhealthy 
  if the  still fails.
  
Thank you!

**Review by Dnouv — CHANGES_REQUESTED** [review]

File: superprofile/initFaunaOnce.sh (hunk 1 above)
Comment (Palanikannan1437) [reply superprofile/initFaunaOnce.sh:30]: Ohhh!! That's a really good catch!⚡️ Thank you so much @Dnouv for the detailed explanation and review, I'll make the necessary changes!

Thank you 😄!

File: superprofile/initFaunaOnce.sh [hunk 2]
```diff
@@ -17,20 +20,33 @@ sleep $waittime
...
+            echo "Process ended with health status of Container: $container_state"
+        fi
     else
-        echo "-- Env variables are already copied, no need to copy over twice 😉 --" 
+        docker exec -it faunadb /bin/sh  /var/log/faunadb/initialize.sh $1
+        if [ -f log/dbkey ] && [ ! -f log/init_key_flag ]; then
+            echo "Copying over secrets to ../app/.env"
+            printf '\nNEXT_PUBLIC_FAUNA_SECRET=' | cat - ./log/dbkey >> ../app/.env &&
+            printf '\nNEXT_PUBLIC_FAUNA_DOMAIN'="http://localhost:8084/graphql" >> ../app/.env
+            touch $DBF &&
```
[the remark this answers is not in the document]
Comment (Palanikannan1437) [reply superprofile/initFaunaOnce.sh]: Sure thing @Dnouv! I've made the necessary changes, thank you 😄!

**Comment by Dnouv** [comment]
LGTM! Thanks for the hard work @Palanikannan1437

**Comment by Palanikannan1437** [comment]
> LGTM! Thanks for the hard work @Palanikannan1437
> 
> cc: @Sing-Li

Thanks @Dnouv!

File: .vscode/settings.json [hunk 3]
```diff
@@ -2,5 +2,6 @@
     "eslint.workingDirectories": [
         "./app",
         "./cms"
-    ]
+    ],
+    "workbench.startupEditor": "none",
```
Comment (Sing-Li) [inline .vscode/settings.json:6]: didn't seem to work :(    I tested it twice.

**Review by Sing-Li — CHANGES_REQUESTED** [review]

File: .vscode/settings.json (hunk 3 above)
Comment (Palanikannan1437) [reply .vscode/settings.json:6]: @Sing-Li I've fixed it now, could you please check now...thank you!

This is actually a bug from VSCode and is still being resolved...I'll be tracking it as well!
https://github.com/microsoft/vscode/issues/58038

And as for the "Gitpod Release Notes" page opening up at start when new user comes in is being tracked and worked over here...though this is just once for new users!
https://github.com/gitpod-io/gitpod/issues/13638
https://github.com/gitpod-io/gitpod/issues/14533

### Detailed Explanation
Due to [this bug](https://github.com/microsoft/vscode/issues/58038) in VSCode of **workspace setting** not overriding the **user settings**, 
If the user manually overrides the settings before ( in your case when you opened Gitpod for the first time..Gitpod automatically sets this value to `"workbench.startupEditor": "welcomePage",` in your **user settings**)

And hence as all this happened before me making changes to the **workspace setting** of `"workbench.startupEditor": "none",`, you'd be experiencing this bug...since **workspace setting** is currently not overriding the **user settings**

SO in conclusion **NEW** gitpod users i.e. without any **user settings** on `"workbench.startupEditor": "..."`, would have the ideal workflow of the website preview and a file like `brandlogo.js` opening side by side!

Comment (Sing-Li) [reply .vscode/settings.json:6]: Amazing.  

Cool.  Tested.  Agreed this is the best we can do for now.
